extensions-contrib/rocketmq-indexing-service/src/main/java/org/apache/druid/indexing/rocketmq/RocketMQRecordSupplier.java

```diff
@@ -0,0 +1,243 @@
+      ObjectMapper sortingMapper
+  )
+  {
+    this(getRocketMQConsumer(sortingMapper, consumerProperties));
+  }
+
+  @VisibleForTesting
+  public RocketMQRecordSupplier(
+      DefaultLitePullConsumer consumer
+  )
+  {
+    this.consumer = consumer;
+  }
+
+  @Override
+  public void assign(Set<StreamPartition<String>> streamPartitions)
+  {
+    this.streamPartitions = streamPartitions;
+    Set<MessageQueue> messageQueues = new HashSet<>();
+    for (StreamPartition<String> streamPartition : streamPartitions) {
+      MessageQueue mq = PartitionUtil.genMessageQueue(streamPartition.getStream(), streamPartition.getPartitionId());
+      messageQueues.add(mq);
+    }
+
+    if (!messageQueues.isEmpty()) {
+      consumer.assign(messageQueues);
+    }
+
+    try {
+      if (!consumer.isRunning()) {
+        consumer.start();
+      }
+    }
+    catch (MQClientException e) {
+      log.error(e.getErrorMessage());
+    }
+    clientInstance = MQClientManager.getInstance().getOrCreateMQClientInstance(consumer);
+  }
+
+  @Override
+  public void seek(StreamPartition<String> partition, Long sequenceNumber)
+  {
+    MessageQueue mq = PartitionUtil.genMessageQueue(partition.getStream(), partition.getPartitionId());
+    try {
+      consumer.seek(mq, sequenceNumber);
+    }
+    catch (MQClientException e) {
+      log.error(e.getErrorMessage());
+      consumer.getOffsetStore().updateOffset(mq, sequenceNumber, false);
+    }
+  }
+
+  @Override
+  public void seekToEarliest(Set<StreamPartition<String>> partitions)
+  {
+    for (StreamPartition<String> partition : partitions) {
+      MessageQueue mq = PartitionUtil.genMessageQueue(partition.getStream(), partition.getPartitionId());
+      try {
+        consumer.seekToBegin(mq);
+      }
+      catch (MQClientException e) {
+        log.error(e.getErrorMessage());
+      }
+    }
+  }
+
+  @Override
+  public void seekToLatest(Set<StreamPartition<String>> partitions)
+  {
+    for (StreamPartition<String> partition : partitions) {
+      MessageQueue mq = PartitionUtil.genMessageQueue(partition.getStream(), partition.getPartitionId());
+      try {
+        consumer.seekToEnd(mq);
+      }
+      catch (MQClientException e) {
+        log.error(e.getErrorMessage());
+      }
+    }
+  }
+
+  @Override
+  public Set<StreamPartition<String>> getAssignment()
+  {
+    if (this.streamPartitions == null) {
+      return new HashSet<StreamPartition<String>>();
+    }
+    return this.streamPartitions;
+  }
+
+  @Nonnull
+  @Override
+  public List<OrderedPartitionableRecord<String, Long, RocketMQRecordEntity>> poll(long timeout)
+  {
+    List<OrderedPartitionableRecord<String, Long, RocketMQRecordEntity>> polledRecords = new ArrayList<>();
+    for (MessageExt messageExt : consumer.poll(timeout)) {
+
+      polledRecords.add(new OrderedPartitionableRecord<>(
+          messageExt.getTopic(),
+          PartitionUtil.genPartition(messageExt.getBrokerName(), messageExt.getQueueId()),
```

Comment:
function call here would create lots of temporary objects. Could we compute the partition id  when paritions are assigned ?

Reply:
BrokerName and QueueID are different in every message. So it must generate one String when get a new message.